File: java/src/main/java/com/tyouqu/database/Dialects.java
package com.tyouqu.database;

public final class Dialects {
    private Dialects() {
    }

    public static SqlDialect forProvider(DatabaseProvider provider) {
        return switch (provider) {
            case SQL_SERVER -> new SqlServerDialect();
            case MYSQL -> new MySqlDialect();
            case POSTGRESQL -> new PostgreSqlDialect();
            case SQLITE -> new SqliteDialect();
        };
    }

    private static final class SqlServerDialect implements SqlDialect {
        public String name() {
            return DatabaseProvider.SQL_SERVER.name();
        }

        public String parameterPrefix() {
            return ":";
        }

        public String quoteIdentifier(String identifier) {
            return "[" + identifier.replace("]", "]]") + "]";
        }

        public String buildPagedSql(String sql, int offset, int pageSize) {
            return trim(sql) + System.lineSeparator() + "offset " + offset + " rows fetch next " + pageSize + " rows only";
        }
    }

    private static final class MySqlDialect implements SqlDialect {
        public String name() {
            return DatabaseProvider.MYSQL.name();
        }

        public String parameterPrefix() {
            return ":";
        }

        public String quoteIdentifier(String identifier) {
            return "`" + identifier.replace("`", "``") + "`";
        }

        public String buildPagedSql(String sql, int offset, int pageSize) {
            return trim(sql) + System.lineSeparator() + "limit " + pageSize + " offset " + offset;
        }
    }

    private static final class PostgreSqlDialect implements SqlDialect {
        public String name() {
            return DatabaseProvider.POSTGRESQL.name();
        }

        public String parameterPrefix() {
            return ":";
        }

        public String quoteIdentifier(String identifier) {
            return "\"" + identifier.replace("\"", "\"\"") + "\"";
        }

        public String buildPagedSql(String sql, int offset, int pageSize) {
            return trim(sql) + System.lineSeparator() + "limit " + pageSize + " offset " + offset;
        }
    }

    private static final class SqliteDialect implements SqlDialect {
        public String name() {
            return DatabaseProvider.SQLITE.name();
        }

        public String parameterPrefix() {
            return ":";
        }

        public String quoteIdentifier(String identifier) {
            return "\"" + identifier.replace("\"", "\"\"") + "\"";
        }

        public String buildPagedSql(String sql, int offset, int pageSize) {
            return trim(sql) + System.lineSeparator() + "limit " + pageSize + " offset " + offset;
        }
    }

    private static String trim(String sql) {
        return sql.trim().replaceFirst(";\\s*$", "");
    }
}

File: python/src/tyouqu_database/dialects.py
from dataclasses import dataclass

from .models import DatabaseProvider


def _trim(sql: str) -> str:
    return sql.strip().removesuffix(";")


@dataclass(frozen=True)
class SqlDialect:
    name: str
    parameter_prefix: str
    quote_start: str
    quote_end: str
    quote_escape: str
    page_style: str

    def quote_identifier(self, identifier: str) -> str:
        escaped = identifier.replace(self.quote_end, self.quote_escape)
        return f"{self.quote_start}{escaped}{self.quote_end}"

    def build_paged_sql(self, sql: str, offset: int, page_size: int) -> str:
        if self.page_style == "sqlserver":
            return f"{_trim(sql)}\noffset {offset} rows fetch next {page_size} rows only"
        return f"{_trim(sql)}\nlimit {page_size} offset {offset}"


def dialect_for_provider(provider: DatabaseProvider) -> SqlDialect:
    if isinstance(provider, str):
        provider = DatabaseProvider(provider.lower())
    if provider == DatabaseProvider.SQL_SERVER:
        return SqlDialect(provider.name, ":", "[", "]", "]]", "sqlserver")
    if provider == DatabaseProvider.MYSQL:
        return SqlDialect(provider.name, ":", "`", "`", "``", "limit")
    if provider in (DatabaseProvider.POSTGRESQL, DatabaseProvider.SQLITE):
        return SqlDialect(provider.name, ":", '"', '"', '""', "limit")
    raise ValueError(f"Unsupported provider: {provider}")
